### PR TITLE
Transitions Callback to Upcall for 2.0. In a seprate PR for review sanity.

### PR DESCRIPTION
**Pull request by phil-levis**
### Pull Request Overview

This pull request changes `kernel::callback::Callback` to `kernel::upcall::Upcall`. It updates the rest of the codebase to use this, and also tries to change all of the comments to distinguish general callbacks from upcalls.

Given that this touches so many files, I suggest that we merge this after the major code review for 2.0.

### Testing Strategy

This pull request was tested by compiling.

### TODO or Help Wanted

I think the terminology is 95% right -- given the scope of these changes, I think that simply fixing the last few when we encounter them will be sufficient.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: chips/litex/src/uart.rs [hunk 1]
```diff
@@ -344,12 +344,12 @@ impl<'a, R: LiteXSoCRegisterConfiguration> uart::Transmit<'a> for LiteXUart<'a,
...
         // If it does not fill the FIFO, an
         // interrupt will _not_ be generated and hence we have to
-        // perform the upcall using a deferred call.
+        // perform the callback using a deferred call.
         //
         // If we fill up the FIFO, an interrupt _will_ be
         // generated. We can transmit the rest using `resume_tx` and
         // directly call the callback there, as we are guaranteed to
-        // be in an upcall.
+        // be in a callback.
```
Comment (hudson-ayers) [inline chips/litex/src/uart.rs:352]: Shouldn't this stay as upcall? for tock's rules about scheduling callbacks, I think what matters is that the callback is only called from within an upcall

File: doc/Syscalls.md [hunk 2]
```diff
@@ -61,13 +61,13 @@ In Tock, a process can be in one of seven states:
...
- in the Running state until it explicitly yields. Callbacks from other kernel
- operations are not delivered to Running processes (i.e. callbacks do not
+ in the Running state until it explicitly yields. Upcalls from other kernel
+ operations are not delivered to Running processes (i.e. upcalls do not
  interrupt processes), rather they are enqueued until the process yields.
  - **Yielded**: Suspended operation. A Yielded process will not be scheduled by
  Tock. Processes often yield while they are waiting for I/O or other operations
  to complete and have no immediately useful work to do. Whenever the kernel issues
- a callback to a Yielded process, the process is transitioned to the Running state.
+ a upcall to a Yielded process, the process is transitioned to the Running state.
```
Comment (hudson-ayers) [inline doc/Syscalls.md:70]:
```suggestion
 an upcall to a Yielded process, the process is transitioned to the Running state.
```

File: doc/Syscalls.md [hunk 3]
```diff
@@ -156,64 +156,70 @@ in the libtock-c syscall code for more information about Yield.
...
+causes the trap handler to execute. The trap handler checks
+`mscratch`, and if the value is nonzero then it contains the stack
+pointer of the kernel and this trap must have happened while the
+system was executing an application. Then, the kernel stack pointer
+from `mscratch` is used to find the pointer to the stored state
+struct, and all process registers are saved. The trap handler also
+saves the process PC from the `mepc` CSR and the `mcause` CSR. It then
+loads the kernel address of where to resume the context switching code
+to `mepc` and calls `mret` to exit the trap handler. Back in the
+context switching code, the kernel restores its registers from its
```
Comment (hudson-ayers) [inline doc/Syscalls.md:192]: This whole diff seems unrelated to this PR?

**Review by hudson-ayers — APPROVED** [review]
LGTM, but a few comments

File: chips/litex/src/uart.rs (hunk 1 above)
Comment (lschuermann) [reply chips/litex/src/uart.rs:352]: I think callback is better here, given that this references an "upcall" from the scheduler, right? That shouldn't be confused with an upcall to userspace. The important thing here is that we're in a different stack frame to be able to safely call `resume_tx`, such that it can enter it's resources again. This is actually the exact issue that confused me in the initial discussion for this in the first place. I suppose saying "upcall from the scheduler" could be a valid option? For reference, this is @alevy's comment which made me call this an "upcall": https://github.com/tock/tock/issues/1736#issuecomment-610015515 :smile:

**Review by lschuermann — COMMENTED** [review]
This looks pretty good, apart from Hudson's comments! It needs a rebase though.

**Review by lschuermann — APPROVED** [review]

**Review by alistair23 — APPROVED** [review]

File: doc/Syscalls.md (hunk 3 above)
Comment (phil-levis) [reply doc/Syscalls.md:192]: It's just a formatting/line wrapping change.
